File: algorithms/sorting/tree-sort/python/tree_sort.py
class Node:
    def __init__(self, key):
        self.left = None
        self.right = None
        self.val = key

def insert(root, key):
    if root is None:
        return Node(key)
    else:
        if key < root.val:
            root.left = insert(root.left, key)
        else:
            root.right = insert(root.right, key)
    return root

def store_sorted(root, arr, index):
    if root is not None:
        index = store_sorted(root.left, arr, index)
        arr[index] = root.val
        index += 1
        index = store_sorted(root.right, arr, index)
    return index

def tree_sort(arr):
    if not arr:
        return arr
        
    root = None
    for x in arr:
        root = insert(root, x)
        
    store_sorted(root, arr, 0)
    return arr

File: algorithms/sorting/tree-sort/swift/TreeSort.swift
class TreeSort {
    private class Node {
        var key: Int
        var left: Node?
        var right: Node?
        
        init(_ key: Int) {
            self.key = key
            self.left = nil
            self.right = nil
        }
    }
    
    static func sort(_ arr: inout [Int]) {
        var root: Node? = nil
        for value in arr {
            root = insert(root, value)
        }
        
        var index = 0
        storeSorted(root, &arr, &index)
    }
    
    private static func insert(_ root: Node?, _ key: Int) -> Node {
        guard let root = root else {
            return Node(key)
        }
        
        if key < root.key {
            root.left = insert(root.left, key)
        } else {
            root.right = insert(root.right, key)
        }
        
        return root
    }
    
    private static func storeSorted(_ root: Node?, _ arr: inout [Int], _ index: inout Int) {
        if let root = root {
            storeSorted(root.left, &arr, &index)
            arr[index] = root.key
            index += 1
            storeSorted(root.right, &arr, &index)
        }
    }
}
